// flow-parser/flow_parser.js
// Generated by js_of_ocaml 4.0.0+git-v0.24.0-11620-gaaac587
(function(a){typeof
globalThis!=="object"&&(this?b():(a.defineProperty(a.prototype,"_T_",{configurable:true,get:b}),_T_));function
b(){var
b=this||self;b.globalThis=b;delete
a.prototype._T_}}(Object));(function(ar){"use strict";var
d2="member_property_expression",DA=8483,Gq=12538,AH="children",e3="predicate_expression",Dz="??",Dy="Identifier",ov=64311,i4=192,Gp=11710,AG=122654,Go=110947,Gn=67591,Q0="!",Q1="directive",SS=163,cZ="block",q5=126553,AF=12735,p_=68096,c3="params",yo=93071,aO=122,Gm=72767,AE=181,g5="for_statement",a8=128,AD="start",pm=43867,R0="_method",Gl=70414,i3=">",e2="catch_body",yn=120121,P6="the end of an expression statement (`;`)",Gk=124907,SR=1027,ou=126558,d1="jsx_fragment",AC=42527,ym="decorators",Gj=82943,Dx=71039,yk=110882,yl=67514,RY=8472,RZ="update",yj=12783,Dw=12438,AB=12352,Gi=8511,AA=42961,kC="method",p9=120713,g4=8191,d0="function_param",Az=67871,cr="throw",Ay=11507,f5="class_extends",Gh=43470,dZ="object_key_literal",Dv=71903,Gg=65437,e1="jsx_child",Gf=43311,q4=119995,yi=67637,ot=68116,yh=66204,Du=65470,SQ="<<=",Ge="e",yg=67391,q3=11631,os=69956,i2="tparams",Gd=66735,Ax=64217,Dt=43697,QZ="Invalid binary/octal ",SP=5761,yf=-43,ye=43255,or="do",Aw=43301,e0="binding_pattern",Ds=120487,g3="jsx_attribute_value_literal",pl="package",eZ="interface_declaration",Gc=72750,Av=119892,QY="tail",QX=-53,eY=111,P5=180,yd=119807,Au=71959,QW=8206,Gb=65613,hs="type",yc=55215,Dr=-42,f4="export_default_declaration_decl",oq=72970,QV="filtered_out",Dq=70416,QU=229,f3="function_this_param",SO="module",pk="try",yb=70143,At=125183,ya=70412,r="@])",g2="binary",SN="infinity",p8="private",Dp=65500,q2="has_unknown_members",eX="pattern_array_rest_element",P4="Property",hQ="implements",Ga=12548,RX=211,f2="if_alternate_statement",F$=124903,Do=43395,jf="src/parser/type_parser.ml",F_=66915,q1=126552,x$=120712,pj=126555,x_=120596,b8="raw",cq=112,f1="class_declaration",eW="statement",Dn=126624,As=71235,f0="meta_property",x8=44002,x9=8467,dY="class_property_value",Dm=8318,dX="optional_call",Ar=43761,hr="kind",fZ="class_identifier",F9=69955,x7=66378,Aq=120512,Dl=68220,a9=110,F8=123583,ku="declare",fY="typeof_member_identifier",eV="catch_clause",Dk=11742,F7=70831,op=8468,dW="for_in_assignment_pattern",P3=-32,fX="object_",x5=43262,x6="mixins",dV="type_param",RW="visit_trailing_comment",Ap=71839,kt="boolean",eU="call",SM="expected *",Ao=43010,x4=241,bE="expression",kI="column",x3=43595,Dj=43258,x2=191456,g1="member_type_identifier",km=117,Di=43754,q0=126544,SL="Assert_failure",x0=66517,x1=42964,fW="enum_number_member",P2="a string",An=65855,Dh=119993,xZ="opaque",QT=870530776,Am=67711,Dg=66994,g0="enum_symbol_body",P1=185,QS=219,p7="filter",Al=43615,p6=126560,F6=19903,ij="get",Df=64316,QR="Fatal error: exception %s\n",p5="exported",SK=">=",bW="return",on="members",oo=256,F5=66962,xY=64279,F4=67829,P0="Enum `",QQ="&&=",eT="object_property",F3=67589,gZ="pattern_object_property",dU="template_literal_element",De=69551,c2=127343600,qZ=70452,gY="class_element",F2="ENOENT",Dd=71131,PZ=200,F0=120137,F1=94098,qY=72349,gX="function_identifier",xX=126543,eS="jsx_attribute_name",Dc=43487,H="@[<2>{ ",SJ="ENOTEMPTY",xW=65908,Db=72191,om=120513,Ak=92909,QP="bound",Da=162,PY=172,ol=120070,dT="enum_number_body",fV="update_expression",dS="spread_element",fU="for_in_left_declaration",pi=64319,ks="%d",C$=12703,qX=11687,SI="@,))@]",FZ=42239,gW="type_cast",xV=42508,eR="class_implements_interface",xU=67640,Aj=605857695,QO="Cygwin",RV="buffer.ml",Ai=124908,SH="handler",xT=66207,Ah=66963,qW=11558,SG="-=",au=113,C_=113775,PX="collect_comments",p4=126540,i1="set",fT="assignment_pattern",bC="right",fS="object_key_identifier",ph=120133,C9="Invalid number ",xS=42963,ok=12539,C7=68023,C8=43798,cO=100,gV="pattern_literal",dR="generic_type",QN="*",Ag=42783,C6=42890,xR=230,qV="else",xP=70851,xQ=69289,PW="the start of a statement",oj="properties",xO=43696,Af=110959,fR="declare_function",qU=120597,gU="object_indexer_property_type",Ae=70492,xN=2048,kl="arguments",J="comments",C5=43042,hn=107,xM=110575,RU=161,FY=67431,oi="line",kr="declaration",bc="static",fQ="pattern_identifier",FX=69958,SF="the",FW="Unix.Unix_error",xL=43814,hq="annot",xK=65786,C3=66303,C4=64967,FU=64255,FV=8584,oh=120655,PV="Stack_overflow",FT=43700,dQ="syntax_opt",RT="/static/",gT="comprehension",C2=253,QM="Not_found",SE="+=",xJ=235,xG=68680,xH=66954,xI=64324,Ad=72966,SD=174,QL=-1053382366,xF="rest",dO="pattern_array_element",dP="jsx_attribute_value_expression",qT=65595,gS="pattern_array_e",SC=243,FS=43711,xE="rmdir",og="symbol",C1=69926,xD=43741,ch="typeParameters",kk="const",PU=1026,QK=149,xC=12341,C0=72847,xB=66993,QJ=202,cX="false",a3=106,xz=120076,xA=186,cW=0x80,Ab=125124,Ac="Fatal error: exception ",pg=67593,CZ=69297,xy=44031,QI=234,QH="all_comments",xx=92927,Aa=68095,bD=8231,gR="object_key_computed",fP="labeled_statement",gQ="function_param_pattern",pf=126590,CX=65481,CY=43442,RS="collect_comments_opt",gP="variable_declarator",jj="_",CW="compare: functional value",CV=67967,je="computed",fO="object_property_type",aN="id",CU=126562,hP=114,RR="comment_bounds",FR=70853,FQ=69247,dN="class_private_field",CT=42237,z$=72329,PT="Invalid_argument",FP=113770,pe=94031,CS=120092,fN="declare_class",z_=67839,FO=72250,PS=249,PR="%ni",CR=92879,SB="prototype",hM="`.",c1=8287,of=65344,z9="&",ci="debugger",fM="type_identifier_reference",SA="Internal Error: Found private field in object props",gO="sequence",dM="call_type_args",QG=238,FN=12348,Sz="++",FM=68863,xv=72001,xw=70084,FL="label",i9=-45,eQ="jsx_opening_attribute",CQ=43583,qS="%F",xu=43784,z8=113791,fL="call_arguments",pd=126503,xt=43743,bV="0",z7=119967,qR=126538,gN="new_",i0=449540197,z6=64109,CP=68466,CO=177983,aP=248,dL="program",ac="@,]@]",FK=68031,dK="function_type",fK="type_",pc=8484,CN=67382,CL=42537,CM=226,FJ=66559,CK=42993,FI=64274,qQ=71236,xs=120069,z5=72105,z3=126570,z4="object",xr=42959,cg="break",fJ="for_of_statement",xq=43695,qP=126551,CJ=66955,p3=126520,xp=66499,kB=1024,xn=67455,xo=43018,QF=198,oe=126522,gM="function_declaration",xm=73064,eP="await",FG=92728,FH=70418,FF=68119,fI="function_rest_param",FE=42653,qO=11703,cL="left",qN=70449,FD=184,gL="declare_type_alias",fH=0xffffff,qM=70302,Sy="/=",QE="|=",FC=55242,FA=126583,FB=124927,xk=124895,xl=72959,CI=65497,RQ="Invalid legacy octal ",hp="typeof",xj="explicit_type",gK="statement_list",CH=65495,fG="class_method",pb=8526,p2=244,Fz=67861,qL=119994,od="enum",QD=2147483647,CG=69762,Sx=208,kq="in",Fy=11702,oc=67638,PQ=", characters ",xi=70753,jd="super",Fx=92783,xh=8304,ob=126504,gJ="import_specifier",z2=68324,xg=101589,CF=67646,fF="expression_or_spread",z1=74879,CE=43792,pa=43260,Fw=93052,QC="{",z0=65574,CD=125258,ji=224,dJ="jsx_element_name_member_expression",kA="instanceof",CB=69599,CC=43560,gI="function_expression",oa=223,xf=72242,CA=11498,Fu=126467,Fv=73112,RP=140,p1=70107,xe=13311,fE="jsx_children",qK=126548,xd=63743,n$=43471,dI="jsx_expression",zZ=69864,Cz=71998,xc=72000,qJ=126591,n_=12592,dH="type_params",Cy=126578,n9=126537,G="{ ",Cx=123627,gH="jsx_spread_attribute",_="@,",Cw=70161,xb=187,o$=126500,eO="label_identifier",Cv=42606,gG="number_literal_type",p0=42999,Ft=64310,QB=-594953737,Fs=122623,qI="hasUnknownMembers",dG="array",RO="^=",eN="enum_string_member",Cu=65536,Fr=65615,hl="void",zY=65135,x=")",Sw=138,zX=70002,kz="let",xa=70271,Ct="nan",o="@[%s =@ ",Cs=194559,Fq=110579,dF="binding_type_identifier",Cr=42735,Sv=0xdfff,bT="/",eM="for_in_statement_lhs",Fo=43503,Fp=8516,Fm=66938,Fn="ENOTDIR",PP="TypeParameterInstantiation",Cp=69749,Cq=65381,Co=83526,jc="number",w$=12447,QA=154,pZ=70286,Fk=72160,Fl=43493,PO=206,eL="enum_member_identifier",o_=70280,kj="function",pY=70162,Cn=255,w_=67702,Cm=66771,w9=70312,Qz="|",zW=93759,RN="End_of_file",w8=43709,ii="new",Cl="Failure",kp="local",w7=101631,o9=8489,pX="with",gF="enum_declaration",n8=218,zV=70457,o8=8488,fD="member",qH=64325,Fj=247,zU=70448,Ck=69967,qG=126535,w6=71934,dE="import_named_specifier",Cj=65312,Fi=126619,fC="type_annotation",RM=0xdc00,Fg=131071,Fh=120770,w5=67002,fA="with_",fB="statement_fork_point",RL="finalizer",w4=12320,PN="elements",dD="literal",zT=68607,w3=8507,o7="each",Su="Sys_error",Fe=123535,Ff=130,eK="bigint_literal_type",zS=64829,pW=11727,w2=120538,eJ="member_private_name",fz="type_alias",Qy="Printexc.handle_uncaught_exception",qF=126556,gE="tagged_template",eI="pattern_object_property_literal_key",Fd=43881,n7=72192,Ci=67826,w0=124910,w1=66511,hk="int_of_string",Ch=43249,y="None",St="FunctionTypeParam",cA="name",wZ=70285,cp=103,wY=120744,dB=12288,dC="intersection_type",Fc=11679,qE=11559,PM="callee",Cg=71295,wX=70018,Cf=11567,Fb=42954,Ss="*-/",bX="predicate",dA="expression_statement",RK="regexp",wW=65479,Qx=132,Ce=11389,bL="optional",PL=-602162310,m="@]",wV=120003,zR=72249,Qw="Unexpected ",Cd=73008,pV="finally",fy="toplevel_statement_list",RJ="end",Fa=178207,Sr="&=",E$=70301,Sq="%Li",Cb=72161,Cc=69746,wT=70460,wU=12799,o6=65535,Ca="loc",zQ=69375,B$=43518,PK=205,B_=65487,gD="while_",B9=183983,eH="typeof_expression",E_=-673950933,E9=42559,Sp="||",zP=124926,wS=55291,gC="jsx_element_name_identifier",c5=8239,n6="mixed",B8=-253313196,zO=11734,pU=67827,wR=68287,E8=119976,So="**",n=" =",o5=888960333,zN=124902,eG="tuple_type",Sn=227,wQ=70726,B7=73111,pT=126602,wP=126529,eF="object_property_value_type",t="%a",RI=", ",Sm="<=",wO=69423,PJ=199,o4=11695,B6=12294,pS=11711,E7=67583,Sl=710,qD=126584,wN=68295,B5=72703,B4="prefix",RH=-80,qC=69415,wM=11492,kH="class",pR=65575,co="continue",zM=65663,RG=2047,n5=68120,zK=71086,zL=19967,c4=782176664,E6=120779,n4=8486,cN=" ",PI="||=",Qv="Undefined_recursive_module",B3=66863,Sk="RestElement",o3=126634,zJ=66377,E5=74751,gB="jsx_element_name_namespaced",wL=43334,wK=66815,cn="typeAnnotation",E4=120126,gA="array_element",qB=64285,Si=189,Sj="**=",L="()",B2=8543,dz="declare_module",eE="export_batch_specifier",RF="%i",RE=">>>=",E3=68029,Sh="importKind",cm="extends",zI=64296,n3=43259,zH=71679,E2=64913,wI=119969,wJ=94175,B1=72440,qA=65141,eD="function_",B0=43071,BZ=42888,E1=69807,bl="variance",hm=123,fx="import_default_specifier",PH=">>>",BY=43764,cM="pattern",BX=71947,E0=70655,jb="consequent",RD=4096,zG=183,EZ=68447,wH=65473,hj=0xff,BW=73648,fw="call_type_arg",gz=8238,EY=68899,wG=93026,ab="@[<2>[",EX=110588,dy="comment",RC=191,gy="switch_case",Qu=175,BV=71942,fv="do_while",i8="constructor",wF=43587,BU=43586,bs="yield",zF=67462,PG="fd ",wE=-61,BT="target",pQ=72272,ki="var",Qt="impltype",qz=70108,kh="0o",zE=119972,BS=92991,o2=70441,qy=8450,wD=120074,EW=66717,gx="interface_type",pP=43880,at="%B",zD=111355,iZ=5760,EV=11630,o1=126499,BR="of",PF=">>",Qs="Popping lex mode from empty stack",o0=120629,ir=108,wC=43002,RB="%=",pO=126539,EU=126502,dx="template_literal",zC="src/parser/statement_parser.ml",zB=": Not a directory",PE="b",zA=67461,BQ=11519,Qr="src/parser/flow_lexer.ml",Qq="Out_of_memory",zz=120570,fu=12287,BP=126534,ET="index out of bounds",BO=73029,qx="_bigarr02",oZ=126571,RA="))",ft="for_statement_init",Sg="supertype",gw="class_property",oY="}",iq="this",gv="declare_module_exports",Qp="@",dw="union_type",cV=0xffff,eC="variance_opt",BN=94032,Rz=222,ES=42124,gu="this_expression",eB="jsx_element",Qo="typeArguments",BM=65019,EQ=125251,ER=64111,EP=8471,eA="typeof_qualified_identifier",BL=70497,Qn="EnumDefaultedMember",fs=8202,wB=66927,cl="switch",EO=69634,gt="unary_expression",zy=71215,Qm=126,wA=67679,BK=65597,Sf=207,wz=120686,n2=72163,zx=67001,zw=42962,BJ=64262,kG=124,dv=65279,BI=126495,Ry=169,EN=71944,PD=-10,pN="alternate",EL=92975,EM=65489,iY=252,EJ=67807,EK=43187,zv=68850,pM="export",wy=66383,Ql="===",fr=".",gs="type_args",Se=147,wx=92159,Qk=0xf0,fq="jsx_element_name",BH=72283,BG=171,ip=116,zu=110587,oX=70279,EH=75075,EI=65338,gr="function_params",ww=126627,Rx=213,qw=73065,wv=71352,n1=119970,zt=70005,BF=12295,n0=120771,EG=71494,EF=11557,BE=42191,Rw="flags",EE=68437,zs=70730,ez="optional_indexed_access",fp="pattern_object_p",zr=42785,ey="nullable_type",ax="value",ED=12343,zp=68415,zq=11694,Qj=221,EC=11726,gq="syntax",EB=119964,PC="&&",BD=68497,EA=73097,hi="null",nZ=126523,Ez=120084,BC=126601,wu=8454,BB="expressions",wt=72144,kg='"',N="(@[",PB=231,zo=170,pL=12448,BA=68786,qv="<",Pz=931,PA="(",Ru=196,Rv=0x800,qu="an identifier",oW=69959,Bz=68799,Sd="leadingComments",Ey=72969,ws=182,By=100351,fo="enum_defaulted_member",Ex=69839,wr=94026,Ew=209,Sc=">>=",wq=131,nY=12336,b6="empty",Px=331416730,Py=204,Ev=70479,Eu=69487,zn=101640,Bx=43123,Pw="([^/]+)",oV=8319,Rt=165,ex="object_type_property_setter",Rs=909,Et=0xf,zm=12591,D=125,wp=92735,Pv="cases",wo=183969,ih="bigint",Rr="Division_by_zero",Bv=67071,Bw=12329,oU=120004,Bu=69414,qt="if",wn=126519,Es="immediately within another function.",Bt=55238,zl=126498,Sb="qualification",wm=66256,F="@ }@]",kf=118,qs=11565,nX=120122,ew="pattern_object_rest_property",Bs=74862,pK="'",zj=-26065557,zk=124911,jh=119,b5=104,fn="assignment",Br=8457,kF="from",Bq=64321,Eq=113817,Er=65629,zi=42655,cK=102,zh=43137,zg=11502,q=";@ ",cz=101,ev="pattern_array_element_pattern",aF="body",eu="jsx_member_expression",wl=65547,et="jsx_attribute_value",es="jsx_namespaced_name",pJ=72967,Ep=126550,ky=254,En=43807,Eo=43738,nW=126589,oT=8455,pI=126628,Em=11670,Pu="*=",oS=120134,er="conditional",Sa=" : flags Open_text and Open_binary are not compatible",nV=119965,wk=69890,ze=72817,zf=164,Bp=43822,qr=69744,Pt="\\\\",Bo=43638,wi=93047,wj="AssignmentPattern",nU=64322,zd=123190,Rq=188,du="object_spread_property_type",zc=70783,El=113663,R$=160,nT=42622,qq=43823,c0="init",jg=109,wh=66503,oR="proto",Bn=74649,fm="optional_member",Ek=40981,wg=120654,e="@ ",dt="enum_boolean_body",eq="export_named_specifier",ep="declare_interface",wf=70451,eo="pattern_object_property_computed_key",pG=-97,pH=120539,qp=64317,we=12543,fl="export_named_declaration_specifier",wd=43359,oP=126530,oQ=72713,Bm=113800,Qi=195,wc=72367,zb=72103,wb=70278,fk="if_consequent_statement",ko=-85,nS=126496,fj="try_catch",go="computed_key",gp="class_",za=173823,en="pattern_object_property_identifier_key",R_="f",em="arrow_function",qo=8485,Ej=126546,el="enum_boolean_member",Bl=94177,kE="delete",y$=232,Rp="blocks",gn="pattern_array_rest_element_pattern",wa=78894,qn=66512,Ei=94111,i7="string",hO="test",v9=69572,v_=66463,v$=66335,v7=72348,v8=73061,ig=":",fi="enum_body",v6=110590,fh="function_this_param_type",Eh=215,v5=77823,Qh="minus",Ro=201,y_=119980,gm="private_name",ds="object_key",ek="function_param_type",Rn="<<",y9=11718,il="as",R9="delegate",cU="true",Eg=67413,oO=70854,Ef=73439,v4=43776,y6=71723,y7=11505,y8=214,Bj=120628,Bk=43513,fg="jsx_attribute_name_namespaced",nQ=120127,nR="Map.bal",oN="any",Ps="@[",R8="camlinternalMod.ml",pF=126559,bK="import",nP=70404,ej="jsx_spread_child",y5=233,Bi=67897,Ee=119974,br=8233,v3=68405,nO=239,Pr="attributes",Qg=173,ff="object_internal_slot_property_type",v2=71351,Ed=242,Bh=67643,nN="shorthand",gl="for_in_statement",v1=126463,Bg=71338,Eb=69445,Ec=65370,v0=73055,y4=167,y3=64911,ei="pattern_object_property_pattern",Pq=212,Rm=197,pE=126579,Bf=64286,Ea="explicitType",y1=67669,y2=43866,Qf="Sys_blocked_io",oM="catch",Be=123197,vZ=64466,vY=65140,vX=73030,Bd=69404,nM="protected",R7=8204,vW=67504,Rl=193,ke=246,y0=43713,pD=120571,fe="array_type",Pp="%u",dr="export_default_declaration",fd="class_expression",R6="quasi",a2="%S",yZ=8525,pC=126515,D$=120485,nL=43519,qm=120745,pB=94178,D_=126588,aC=127,yY=66855,Rk="@{",Rj="visit_leading_comment",vV=67742,Ri=" : flags Open_rdonly and Open_wronly are not compatible",D9=120144,pA="returnType",ik=-744106340,ja=240,gk="-",ql=8469,hN="async",pz=126521,D8=72095,nK=216,Rh=" : file already exists",vT=178205,vU=8449,oL=94179,Bc=42774,qk="case",Bb=66965,vS=66431,Qe=190,dq="declare_export_declaration",kx="targs",gj="type_identifier",yX=64284,Ba=43013,qj=43815,gi="function_body_any",yW=66966,oK=120687,D7=66939,yV=66978,Rg=168,py="public",vR=68115,yU=43712,px=65598,oJ=126547,D6=110591,dp="indexed_access",Po=12520,bU="interface",is=-46,dm="string_literal_type",dn="import_namespace_specifier",vQ=120132,oI=11735,A$=67505,qi=119893,oH="bool",kD=1000,cI="default",yT=236,i="",vP="exportKind",Pn="trailingComments",oG="^",A_=71983,D5=8348,yS=66977,A9=65594,fb="logical",fc="jsx_member_expression_identifier",D4=210,R5="cooked",gh="for_of_left_declaration",iX=0x3f,yR=72202,b7="argument",D3=12442,vO=43645,nJ=120085,D2=42539,oF=126468,R4=166,Rf="Match_failure",vN=68191,bq="src/parser/flow_ast.ml",qh=11647,gg="declare_variable",ho="+",vM=71127,nI=120145,dl="declare_export_declaration_decl",pw=64318,R3=179,eh="class_implements",Pm="!=",Qd="inexact",R2="%li",Re=237,kw="a",oE=73062,A8=178,fa=65278,eg="function_rest_param_type",D1=77711,yP=70066,yQ=43714,Rd=-696510241,oD=70480,yO=69748,A7=113788,vL=94207,Pl="\r\n",dk="class_body",yM=126651,yN=68735,D0=43273,qg=119996,pv=67644,Rc=0xe0,ef="catch_clause_pattern",e$="boolean_literal_type",nH=126554,oC=126557,A6=113807,pu=126536,Rb="%",i6="property",DZ=71956,Pk="#",yL=123213,kn="meta",ee="for_of_assignment_pattern",dj="if_statement",vK=66421,A5=8505,vJ=225,kd=250,yK=100343,pt="Literal",DY=42887,io=115,Ra=";",yJ=1255,A4="=",DW=126566,DX=93823,gf="opaque_type",Q$="!==",e_="jsx_attribute",ge="type_annotation_hint",cY=0x8000,A3=73727,Q9="range",Q_=245,DV="jsError",ps=70006,DU=43492,pr="@]}",z="(Some ",DT=8477,Qc=129,A2=71487,qf=126564,Q8="\n",DS=126514,vI=70080,di="generic_identifier_type",yI=66811,e9="typeof_identifier",Q7="~",DR=65007,gd="pattern_object_rest_property_pattern",vH=194,Qb=1039100673,yH=66461,DQ=70319,pq=11719,DP=72271,a7=-48,ed="enum_string_body",DO=70461,gc="export_named_declaration",vF=110930,vG=92862,Qa="??=",yG=70440,oB="while",bk="camlinternalFormat.ml",vE=43782,P$=203,A1=173791,DN=11263,oA=1114111,yF=42969,oz=70750,ec="jsx_identifier",DM=70105,vD=43014,A0=11564,eb="typeof_type",Q6="EEXIST",vC=64847,vA=71167,vB=42511,yE=72712,AZ=92995,yD=43704,i$=121,ea="object_call_property_type",vz=64433,kv="operator",qe=68296,d$="class_decorator",e7=120,e8="for_of_statement_lhs",DL=11623,yC=67004,vy=71999,yA=70708,yB=512,AY=110927,yz=71423,Q5=0x7ff0,vw=93951,vx=12292,gb="object_type",nG="types",DK=110580,P_=177,vv=126633,yy=12686,e6=8286,Pj=144,AX=73647,Q4=228,P9="tokens",oy=70855,im="0x",vu=70366,AW="\n",dh="variable_declaration",AV=65276,ox=119981,vt=71945,AU=43887,ck=105,DJ=8335,yx=123565,yw=69505,AT=70187,d_="jsx_attribute_name_identifier",f$="source",ga="pattern_object_property_key",qd=65548,DI=66175,yv=92766,d9="pattern_assignment_pattern",f_="object_type_property_getter",pp=8305,cj="generator",nF="for",P7="PropertyDefinition",P8="--",bj=-36,AS="mkdir",AR=68223,d8="generic_qualified_identifier_type",yu=11686,d7="jsx_closing_element",DH=43790,qc=": No such file or directory",AQ=69687,yt=66348,po=72162,vs=43388,vr=72768,AP=68351,f="<2>",qb=64297,vq=125259,ys=220,M=",@ ",Q3="win32",nE=70281,dg="member_property_identifier",AN=68149,AO=68111,vp=71450,DG=43009,d6="member_property",vo=73458,cJ="identifier",vn=67423,vm=66775,DF=110951,Q2="Internal Error: Found object private prop",f9="super_expression",d5="jsx_opening_element",AM=177976,f8="variable_declarator_pattern",df="pattern_expression",e5="jsx_member_expression_object",DE=68252,yr=77808,i_=-835925911,e4="import_declaration",DD=55203,Ph="Pervasives.do_at_exit",Pi="utf8",cy="key",yq=43702,d4="spread_property",pn=126563,vk=863850040,vl=70106,nD=67592,f7="function_expression_or_method",DC=71958,f6="for_init_declaration",vj=71955,nC=123214,AL=68479,Pg="==",AK=43019,vi=123180,nB=217,i5="specifiers",d3="function_body",AJ=69622,qa=8487,vh=43641,Pf="Unexpected token `",Pe="v",yp=123135,AI=69295,p$=120093,DB=8521,ow=43642,R1=176;function
bGA(d,b,e,c,f){if(c<=b)for(var
a=1;a<=f;a++)e[c+a]=d[b+a];else
for(var
a=f;a>=1;a--)e[c+a]=d[b+a];return 0}function
bGB(a){var
c=[0];while(a!==0){var
d=a[1];for(var
b=1;b<d.length;b++)c.push(d[b]);a=a[2]}return c}function
bGE(e,f,d){var
a=new
Array(d+1);a[0]=0;for(var
b=1,c=f+1;b<=d;b++,c++)a[b]=e[c];return a}function
GH(c,b,a){var
d=String.fromCharCode;if(b==0&&a<=RD&&a==c.length)return d.apply(null,c);var
e=i;for(;0<a;b+=kB,a-=kB)e+=d.apply(null,c.slice(b,b+Math.min(a,kB)));return e}function
q7(b){if(ar.Uint8Array)var
c=new(ar.Uint8Array)(b.l);else
var
c=new
Array(b.l);var
e=b.c,d=e.length,a=0;for(;a<d;a++)c[a]=e.charCodeAt(a);for(d=b.l;a<d;a++)c[a]=0;b.c=c;b.t=4;return c}function
hS(d,e,b,f,c){if(c==0)return 0;if(f==0&&(c>=b.l||b.t==2&&c>=b.c.length)){b.c=d.t==4?GH(d.c,e,c):e==0&&d.c.length==c?d.c:d.c.substr(e,c);b.t=b.c.length==b.l?0:2}else
if(b.t==2&&f==b.c.length){b.c+=d.t==4?GH(d.c,e,c):e==0&&d.c.length==c?d.c:d.c.substr(e,c);b.t=b.c.length==b.l?0:2}else{if(b.t!=4)q7(b);var
g=d.c,h=b.c;if(d.t==4)if(f<=e)for(var
a=0;a<c;a++)h[f+a]=g[e+a];else
for(var
a=c-1;a>=0;a--)h[f+a]=g[e+a];else{var
i=Math.min(c,g.length-e);for(var
a=0;a<i;a++)h[f+a]=g.charCodeAt(e+a);for(;a<c;a++)h[f+a]=0}}return 0}function
g8(a){return a}function
hT(a,b,c,d,e){hS(g8(a),b,c,d,e);return 0}function
bHf(b,a){throw[0,b,a]}function
jp(b,a){if(b==0)return i;if(a.repeat)return a.repeat(b);var
d=i,c=0;for(;;){if(b&1)d+=a;b>>=1;if(b==0)return d;a+=a;c++;if(c==9)a.slice(0,1)}}function
jl(a){if(a.t==2)a.c+=jp(a.l-a.c.length,"\0");else
a.c=GH(a.c,0,a.c.length);a.t=0}function
To(a){if(a.length<24){for(var
b=0;b<a.length;b++)if(a.charCodeAt(b)>aC)return false;return true}else
return!/[^\x00-\x7f]/.test(a)}function
GJ(e){for(var
k=i,c=i,g,f,h,a,b=0,j=e.length;b<j;b++){f=e.charCodeAt(b);if(f<cW){for(var
d=b+1;d<j&&(f=e.charCodeAt(d))<cW;d++);if(d-b>yB){c.substr(0,1);k+=c;c=i;k+=e.slice(b,d)}else
c+=e.slice(b,d);if(d==j)break;b=d}a=1;if(++b<j&&((h=e.charCodeAt(b))&-64)==a8){g=h+(f<<6);if(f<Rc){a=g-0x3080;if(a<cW)a=1}else{a=2;if(++b<j&&((h=e.charCodeAt(b))&-64)==a8){g=h+(g<<6);if(f<Qk){a=g-0xe2080;if(a<Rv||a>=0xd7ff&&a<0xe000)a=2}else{a=3;if(++b<j&&((h=e.charCodeAt(b))&-64)==a8&&f<0xf5){a=h-0x3c82080+(g<<6);if(a<0x10000||a>0x10ffff)a=3}}}}}if(a<4){b-=a;c+="\ufffd"}else
if(a>cV)c+=String.fromCharCode(0xd7c0+(a>>10),RM+(a&0x3FF));else
c+=String.fromCharCode(a);if(c.length>kB){c.substr(0,1);k+=c;c=i}}return k+c}function
g6(c,a,b){this.t=c;this.c=a;this.l=b}g6.prototype.toString=function(){switch(this.t){case
9:return this.c;default:jl(this);case
0:if(To(this.c)){this.t=9;return this.c}this.t=8;case
8:return this.c}};g6.prototype.toUtf16=function(){var
a=this.toString();if(this.t==9)return a;return GJ(a)};g6.prototype.slice=function(){var
a=this.t==4?this.c.slice():this.c;return new
g6(this.t,a,this.l)};function
S2(a){return new
g6(0,a,a.length)}function
a(a){return S2(a)}function
GF(c,b){bHf(c,a(b))}var
aX=[0];function
bm(a){GF(aX.Invalid_argument,a)}function
S0(){bm(ET)}function
aG(a,c,b){b&=hj;if(a.t!=4){if(c==a.c.length){a.c+=String.fromCharCode(b);if(c+1==a.l)a.t=0;return 0}q7(a)}a.c[c]=b;return 0}function
it(b,a,c){if(a>>>0>=b.l)S0();return aG(b,a,c)}function
bM(a,b){switch(a.t&6){default:if(b>=a.c.length)return 0;case
0:return a.c.charCodeAt(b);case
4:return a.c[b]}}function
ht(c,a){if(c.fun)return ht(c.fun,a);if(typeof
c!=="function")return c;var
b=c.length|0;if(b===0)return c.apply(null,a);var
e=a.length|0,d=b-e|0;if(d==0)return c.apply(null,a);else
if(d<0)return ht(c.apply(null,a.slice(0,b)),a.slice(b));else
return function(){var
e=arguments.length==0?1:arguments.length,d=new
Array(a.length+e);for(var
b=0;b<a.length;b++)d[b]=a[b];for(var
b=0;b<arguments.length;b++)d[a.length+b]=arguments[b];return ht(c,d)}}function
kJ(){bm(ET)}function
a4(a,b){if(b>>>0>=a.length-1)kJ();return a}function
bGL(a){if(isFinite(a)){if(Math.abs(a)>=2.2250738585072014e-308)return 0;if(a!=0)return 1;return 2}return isNaN(a)?4:3}function
hw(a){a.t&6&&jl(a);return a.c}var
bHp=Math.log2&&Math.log2(1.1235582092889474E+307)==1020;function
bHo(a){if(bHp)return Math.floor(Math.log2(a));var
b=0;if(a==0)return-Infinity;if(a>=1)while(a>=2){a/=2;b++}else
while(a<1){a*=2;b--}return b}function
Gv(c){var
a=new(ar.Float32Array)(1);a[0]=c;var
b=new(ar.Int32Array)(a.buffer);return b[0]|0}var
Tb=Math.pow(2,-24);function
Ti(a){throw a}function
Tj(){Ti(aX.Division_by_zero)}function
aj(b,c,a){this.lo=b&fH;this.mi=c&fH;this.hi=a&cV}aj.prototype.caml_custom="_j";aj.prototype.copy=function(){return new
aj(this.lo,this.mi,this.hi)};aj.prototype.ucompare=function(a){if(this.hi>a.hi)return 1;if(this.hi<a.hi)return-1;if(this.mi>a.mi)return 1;if(this.mi<a.mi)return-1;if(this.lo>a.lo)return 1;if(this.lo<a.lo)return-1;return 0};aj.prototype.compare=function(a){var
b=this.hi<<16,c=a.hi<<16;if(b>c)return 1;if(b<c)return-1;if(this.mi>a.mi)return 1;if(this.mi<a.mi)return-1;if(this.lo>a.lo)return 1;if(this.lo<a.lo)return-1;return 0};aj.prototype.neg=function(){var
a=-this.lo,b=-this.mi+(a>>24),c=-this.hi+(b>>24);return new
aj(a,b,c)};aj.prototype.add=function(a){var
b=this.lo+a.lo,c=this.mi+a.mi+(b>>24),d=this.hi+a.hi+(c>>24);return new
aj(b,c,d)};aj.prototype.sub=function(a){var
b=this.lo-a.lo,c=this.mi-a.mi+(b>>24),d=this.hi-a.hi+(c>>24);return new
aj(b,c,d)};aj.prototype.mul=function(a){var
b=this.lo*a.lo,c=(b*Tb|0)+this.mi*a.lo+this.lo*a.mi,d=(c*Tb|0)+this.hi*a.lo+this.mi*a.mi+this.lo*a.hi;return new
aj(b,c,d)};aj.prototype.isZero=function(){return(this.lo|this.mi|this.hi)==0};aj.prototype.isNeg=function(){return this.hi<<16<0};aj.prototype.and=function(a){return new
aj(this.lo&a.lo,this.mi&a.mi,this.hi&a.hi)};aj.prototype.or=function(a){return new
aj(this.lo|a.lo,this.mi|a.mi,this.hi|a.hi)};aj.prototype.xor=function(a){return new
aj(this.lo^a.lo,this.mi^a.mi,this.hi^a.hi)};aj.prototype.shift_left=function(a){a=a&63;if(a==0)return this;if(a<24)return new
aj(this.lo<<a,this.mi<<a|this.lo>>24-a,this.hi<<a|this.mi>>24-a);if(a<48)return new
aj(0,this.lo<<a-24,this.mi<<a-24|this.lo>>48-a);return new
aj(0,0,this.lo<<a-48)};aj.prototype.shift_right_unsigned=function(a){a=a&63;if(a==0)return this;if(a<24)return new
aj(this.lo>>a|this.mi<<24-a,this.mi>>a|this.hi<<24-a,this.hi>>a);if(a<48)return new
aj(this.mi>>a-24|this.hi<<48-a,this.hi>>a-24,0);return new
aj(this.hi>>a-48,0,0)};aj.prototype.shift_right=function(a){a=a&63;if(a==0)return this;var
c=this.hi<<16>>16;if(a<24)return new
aj(this.lo>>a|this.mi<<24-a,this.mi>>a|c<<24-a,this.hi<<16>>a>>>16);var
b=this.hi<<16>>31;if(a<48)return new
aj(this.mi>>a-24|this.hi<<48-a,this.hi<<16>>a-24>>16,b&cV);return new
aj(this.hi<<16>>a-32,b,b)};aj.prototype.lsl1=function(){this.hi=this.hi<<1|this.mi>>23;this.mi=(this.mi<<1|this.lo>>23)&fH;this.lo=this.lo<<1&fH};aj.prototype.lsr1=function(){this.lo=(this.lo>>>1|this.mi<<23)&fH;this.mi=(this.mi>>>1|this.hi<<23)&fH;this.hi=this.hi>>>1};aj.prototype.udivmod=function(e){var
c=0,b=this.copy(),a=e.copy(),d=new
aj(0,0,0);while(b.ucompare(a)>0){c++;a.lsl1()}while(c>=0){c--;d.lsl1();if(b.ucompare(a)>=0){d.lo++;b=b.sub(a)}a.lsr1()}return{quotient:d,modulus:b}};aj.prototype.div=function(a){var
b=this;if(a.isZero())Tj();var
d=b.hi^a.hi;if(b.hi&cY)b=b.neg();if(a.hi&cY)a=a.neg();var
c=b.udivmod(a).quotient;if(d&cY)c=c.neg();return c};aj.prototype.mod=function(b){var
a=this;if(b.isZero())Tj();var
d=a.hi;if(a.hi&cY)a=a.neg();if(b.hi&cY)b=b.neg();var
c=a.udivmod(b).modulus;if(d&cY)c=c.neg();return c};aj.prototype.toInt=function(){return this.lo|this.mi<<24};aj.prototype.toFloat=function(){return(this.hi<<16)*Math.pow(2,32)+this.mi*Math.pow(2,24)+this.lo};aj.prototype.toArray=function(){return[this.hi>>8,this.hi&hj,this.mi>>16,this.mi>>8&hj,this.mi&hj,this.lo>>16,this.lo>>8&hj,this.lo&hj]};aj.prototype.lo32=function(){return this.lo|(this.mi&hj)<<24};aj.prototype.hi32=function(){return this.mi>>>8&cV|this.hi<<16};function
q_(b,c,a){return new
aj(b,c,a)}function
q9(a){if(!isFinite(a)){if(isNaN(a))return q_(1,0,Q5);return a>0?q_(0,0,Q5):q_(0,0,0xfff0)}var
f=a==0&&1/a==-Infinity?cY:a>=0?0:cY;if(f)a=-a;var
b=bHo(a)+1023;if(b<=0){b=0;a/=Math.pow(2,-PU)}else{a/=Math.pow(2,b-SR);if(a<16){a*=2;b-=1}if(b==0)a/=2}var
d=Math.pow(2,24),c=a|0;a=(a-c)*d;var
e=a|0;a=(a-e)*d;var
g=a|0;c=c&Et|f|b<<4;return q_(g,e,c)}function
kN(a){return a.toArray()}function
SZ(c,b,g){c.write(32,b.dims.length);c.write(32,b.kind|b.layout<<8);if(b.caml_custom==qx)for(var
a=0;a<b.dims.length;a++)if(b.dims[a]<cV)c.write(16,b.dims[a]);else{c.write(16,cV);c.write(32,0);c.write(32,b.dims[a])}else
for(var
a=0;a<b.dims.length;a++)c.write(32,b.dims[a]);switch(b.kind){case
2:case
3:case
12:for(var
a=0;a<b.data.length;a++)c.write(8,b.data[a]);break;case
4:case
5:for(var
a=0;a<b.data.length;a++)c.write(16,b.data[a]);break;case
6:for(var
a=0;a<b.data.length;a++)c.write(32,b.data[a]);break;case
8:case
9:c.write(8,0);for(var
a=0;a<b.data.length;a++)c.write(32,b.data[a]);break;case
7:for(var
a=0;a<b.data.length/2;a++){var
e=kN(b.get(a));for(var
d=0;d<8;d++)c.write(8,e[d])}break;case
1:for(var
a=0;a<b.data.length;a++){var
e=kN(q9(b.get(a)));for(var
d=0;d<8;d++)c.write(8,e[d])}break;case
0:for(var
a=0;a<b.data.length;a++){var
e=Gv(b.get(a));c.write(32,e)}break;case
10:for(var
a=0;a<b.data.length/2;a++){var
d=b.get(a);c.write(32,Gv(d[1]));c.write(32,Gv(d[2]))}break;case
11:for(var
a=0;a<b.data.length/2;a++){var
f=b.get(a),e=kN(q9(f[1]));for(var
d=0;d<8;d++)c.write(8,e[d]);var
e=kN(q9(f[2]));for(var
d=0;d<8;d++)c.write(8,e[d])}break}g[0]=(4+b.dims.length)*4;g[1]=(4+b.dims.length)*8}function
SX(a){switch(a){case
7:case
10:case
11:return 2;default:return 1}}function
bGF(c,e){var
b=ar,a;switch(c){case
0:a=b.Float32Array;break;case
1:a=b.Float64Array;break;case
2:a=b.Int8Array;break;case
3:a=b.Uint8Array;break;case
4:a=b.Int16Array;break;case
5:a=b.Uint16Array;break;case
6:a=b.Int32Array;break;case
7:a=b.Int32Array;break;case
8:a=b.Int32Array;break;case
9:a=b.Int32Array;break;case
10:a=b.Float32Array;break;case
11:a=b.Float64Array;break;case
12:a=b.Uint8Array;break}if(!a)bm("Bigarray.create: unsupported kind");var
d=new
a(e*SX(c));return d}function
Gw(c){var
a=new(ar.Int32Array)(1);a[0]=c;var
b=new(ar.Float32Array)(a.buffer);return b[0]}function
kM(a){return new
aj(a[7]<<0|a[6]<<8|a[5]<<16,a[4]<<0|a[3]<<8|a[2]<<16,a[1]<<0|a[0]<<8)}function
Gx(d){var
f=d.lo,g=d.mi,b=d.hi,c=(b&0x7fff)>>4;if(c==RG)return(f|g|b&Et)==0?b&cY?-Infinity:Infinity:NaN;var
e=Math.pow(2,-24),a=(f*e+g)*e+(b&Et);if(c>0){a+=16;a*=Math.pow(2,c-SR)}else
a*=Math.pow(2,-PU);if(b&cY)a=-a;return a}function
Gr(b){var
d=b.length,c=1;for(var
a=0;a<d;a++){if(b[a]<0)bm("Bigarray.create: negative dimension");c=c*b[a]}return c}function
bGU(b,a){return new
aj(b&fH,b>>>24&hj|(a&cV)<<8,a>>>16&cV)}function
Gy(a){return a.hi32()}function
Gz(a){return a.lo32()}var
bGG=qx;function
hR(c,d,b,a){this.kind=c;this.layout=d;this.dims=b;this.data=a}hR.prototype.caml_custom=bGG;hR.prototype.offset=function(b){var
c=0;if(typeof
b==="number")b=[b];if(!(b
instanceof
Array))bm("bigarray.js: invalid offset");if(this.dims.length!=b.length)bm("Bigarray.get/set: bad number of dimensions");if(this.layout==0)for(var
a=0;a<this.dims.length;a++){if(b[a]<0||b[a]>=this.dims[a])kJ();c=c*this.dims[a]+b[a]}else
for(var
a=this.dims.length-1;a>=0;a--){if(b[a]<1||b[a]>this.dims[a])kJ();c=c*this.dims[a]+(b[a]-1)}return c};hR.prototype.get=function(a){switch(this.kind){case
7:var
d=this.data[a*2+0],b=this.data[a*2+1];return bGU(d,b);case
10:case
11:var
e=this.data[a*2+0],c=this.data[a*2+1];return[ky,e,c];default:return this.data[a]}};hR.prototype.set=function(a,b){switch(this.kind){case
7:this.data[a*2+0]=Gz(b);this.data[a*2+1]=Gy(b);break;case
10:case
11:this.data[a*2+0]=b[1];this.data[a*2+1]=b[2];break;default:this.data[a]=b;break}return 0};hR.prototype.fill=function(b){switch(this.kind){case
7:var
c=Gz(b),e=Gy(b);if(c==e)this.data.fill(c);else
for(var
a=0;a<this.data.length;a++)this.data[a]=a%2==0?c:e;break;case
10:case
11:var
d=b[1],f=b[2];if(d==f)this.data.fill(d);else
for(var
a=0;a<this.data.length;a++)this.data[a]=a%2==0?d:f;break;default:this.data.fill(b);break}};hR.prototype.compare=function(b,g){if(this.layout!=b.layout||this.kind!=b.kind){var
e=this.kind|this.layout<<8,f=b.kind|b.layout<<8;return f-e}if(this.dims.length!=b.dims.length)return b.dims.length-this.dims.length;for(var
a=0;a<this.dims.length;a++)if(this.dims[a]!=b.dims[a])return this.dims[a]<b.dims[a]?-1:1;switch(this.kind){case
0:case
1:case
10:case
11:var
c,d;for(var
a=0;a<this.data.length;a++){c=this.data[a];d=b.data[a];if(c<d)return-1;if(c>d)return 1;if(c!=d){if(!g)return NaN;if(c==c)return 1;if(d==d)return-1}}break;case
7:for(var
a=0;a<this.data.length;a+=2){if(this.data[a+1]<b.data[a+1])return-1;if(this.data[a+1]>b.data[a+1])return 1;if(this.data[a]>>>0<b.data[a]>>>0)return-1;if(this.data[a]>>>0>b.data[a]>>>0)return 1}break;case
2:case
3:case
4:case
5:case
6:case
8:case
9:case
12:for(var
a=0;a<this.data.length;a++){if(this.data[a]<b.data[a])return-1;if(this.data[a]>b.data[a])return 1}break}return 0};function
jk(c,d,b,a){this.kind=c;this.layout=d;this.dims=b;this.data=a}jk.prototype=new
hR();jk.prototype.offset=function(a){if(typeof
a!=="number")if(a
instanceof
Array&&a.length==1)a=a[0];else
bm("Ml_Bigarray_c_1_1.offset");if(a<0||a>=this.dims[0])kJ();return a};jk.prototype.get=function(a){return this.data[a]};jk.prototype.set=function(a,b){this.data[a]=b;return 0};jk.prototype.fill=function(a){this.data.fill(a);return 0};function
SV(c,d,a,b){var
e=SX(c);if(Gr(a)*e!=b.length)bm("length doesn't match dims");if(d==0&&a.length==1&&e==1)return new
jk(c,d,a,b);return new
hR(c,d,a,b)}function
bZ(b){if(!aX.Failure)aX.Failure=[aP,a(Cl),-3];GF(aX.Failure,b)}function
SW(b,v,r){var
i=b.read32s();if(i<0||i>16)bZ("input_value: wrong number of bigarray dimensions");var
p=b.read32s(),j=p&hj,o=p>>8&1,h=[];if(r==qx)for(var
a=0;a<i;a++){var
n=b.read16u();if(n==cV){var
t=b.read32u(),u=b.read32u();if(t!=0)bZ("input_value: bigarray dimension overflow in 32bit");n=u}h.push(n)}else
for(var
a=0;a<i;a++)h.push(b.read32u());var
d=Gr(h),f=bGF(j,d),g=SV(j,o,h,f);switch(j){case
2:for(var
a=0;a<d;a++)f[a]=b.read8s();break;case
3:case
12:for(var
a=0;a<d;a++)f[a]=b.read8u();break;case
4:for(var
a=0;a<d;a++)f[a]=b.read16s();break;case
5:for(var
a=0;a<d;a++)f[a]=b.read16u();break;case
6:for(var
a=0;a<d;a++)f[a]=b.read32s();break;case
8:case
9:var
s=b.read8u();if(s)bZ("input_value: cannot read bigarray with 64-bit OCaml ints");for(var
a=0;a<d;a++)f[a]=b.read32s();break;case
7:var
e=new
Array(8);for(var
a=0;a<d;a++){for(var
c=0;c<8;c++)e[c]=b.read8u();var
q=kM(e);g.set(a,q)}break;case
1:var
e=new
Array(8);for(var
a=0;a<d;a++){for(var
c=0;c<8;c++)e[c]=b.read8u();var
k=Gx(kM(e));g.set(a,k)}break;case
0:for(var
a=0;a<d;a++){var
k=Gw(b.read32s());g.set(a,k)}break;case
10:for(var
a=0;a<d;a++){var
m=Gw(b.read32s()),l=Gw(b.read32s());g.set(a,[ky,m,l])}break;case
11:var
e=new
Array(8);for(var
a=0;a<d;a++){for(var
c=0;c<8;c++)e[c]=b.read8u();var
m=Gx(kM(e));for(var
c=0;c<8;c++)e[c]=b.read8u();var
l=Gx(kM(e));g.set(a,[ky,m,l])}break}v[0]=(4+i)*4;return SV(j,o,h,f)}function
SU(a,b,c){return a.compare(b,c)}function
Te(a,b){return Math.imul(a,b)}function
hv(b,a){a=Te(a,0xcc9e2d51|0);a=a<<15|a>>>32-15;a=Te(a,0x1b873593);b^=a;b=b<<13|b>>>32-13;return(b+(b<<2)|0)+(0xe6546b64|0)|0}function
bGP(a,b){a=hv(a,Gz(b));a=hv(a,Gy(b));return a}function
S$(a,b){return bGP(a,q9(b))}function
SY(c){var
b=Gr(c.dims),d=0;switch(c.kind){case
2:case
3:case
12:if(b>oo)b=oo;var
e=0,a=0;for(a=0;a+4<=c.data.length;a+=4){e=c.data[a+0]|c.data[a+1]<<8|c.data[a+2]<<16|c.data[a+3]<<24;d=hv(d,e)}e=0;switch(b&3){case
3:e=c.data[a+2]<<16;case
2:e|=c.data[a+1]<<8;case
1:e|=c.data[a+0];d=hv(d,e)}break;case
4:case
5:if(b>a8)b=a8;var
e=0,a=0;for(a=0;a+2<=c.data.length;a+=2){e=c.data[a+0]|c.data[a+1]<<16;d=hv(d,e)}if((b&1)!=0)d=hv(d,c.data[a]);break;case
6:if(b>64)b=64;for(var
a=0;a<b;a++)d=hv(d,c.data[a]);break;case
8:case
9:if(b>64)b=64;for(var
a=0;a<b;a++)d=hv(d,c.data[a]);break;case
7:if(b>32)b=32;b*=2;for(var
a=0;a<b;a++)d=hv(d,c.data[a]);break;case
10:b*=2;case
0:if(b>64)b=64;for(var
a=0;a<b;a++)d=S$(d,c.data[a]);break;case
11:b*=2;case
1:if(b>32)b=32;for(var
a=0;a<b;a++)d=S$(d,c.data[a]);break}return d}function
bGR(a,b){b[0]=4;return a.read32s()}function
bHb(a,b){switch(a.read8u()){case
1:b[0]=4;return a.read32s();case
2:bZ("input_value: native integer value too large");default:bZ("input_value: ill-formed native integer")}}function
bG2(c,d){var
b=new
Array(8);for(var
a=0;a<8;a++)b[a]=c.read8u();d[0]=8;return kM(b)}function
bGZ(e,d,b){var
c=kN(d);for(var
a=0;a<8;a++)e.write(8,c[a]);b[0]=8;b[1]=8}function
bGT(a,b,c){return a.compare(b)}function
bGW(a){return a.lo32()^a.hi32()}var
S6={"_j":{deserialize:bG2,serialize:bGZ,fixed_length:8,compare:bGT,hash:bGW},"_i":{deserialize:bGR,fixed_length:4},"_n":{deserialize:bHb,fixed_length:4},"_bigarray":{deserialize:function(a,b){return SW(a,b,"_bigarray")},serialize:SZ,compare:SU,hash:SY},"_bigarr02":{deserialize:function(a,b){return SW(a,b,qx)},serialize:SZ,compare:SU,hash:SY}};function
Gs(a){return S6[a.caml_custom]&&S6[a.caml_custom].compare}function
S4(f,c,d,e){var
b=Gs(c);if(b){var
a=d>0?b(c,f,e):b(f,c,e);if(e&&a!=a)return d;if(+a!=+a)return+a;if((a|0)!=0)return a|0}return d}function
ra(a){return a
instanceof
g6}function
GB(a){return ra(a)}function
S5(a){if(typeof
a==="number")return kD;else
if(ra(a))return iY;else
if(GB(a))return 1252;else
if(a
instanceof
Array&&a[0]===a[0]>>>0&&a[0]<=Cn){var
b=a[0]|0;return b==ky?0:b}else
if(a
instanceof
String)return Po;else
if(typeof
a=="string")return Po;else
if(a
instanceof
Number)return kD;else
if(a&&a.caml_custom)return yJ;else
if(a&&a.compare)return 1256;else
if(typeof
a=="function")return 1247;else
if(typeof
a=="symbol")return 1251;return 1001}function
g9(a,b){if(a<b)return-1;if(a==b)return 0;return 1}function
S1(a,b){a.t&6&&jl(a);b.t&6&&jl(b);return a.c<b.c?-1:a.c>b.c?1:0}function
X(a,b){return S1(a,b)}function
q6(a,b,d){var
e=[];for(;;){if(!(d&&a===b)){var
f=S5(a);if(f==kd){a=a[1];continue}var
g=S5(b);if(g==kd){b=b[1];continue}if(f!==g){if(f==kD){if(g==yJ)return S4(a,b,-1,d);return-1}if(g==kD){if(f==yJ)return S4(b,a,1,d);return 1}return f<g?-1:1}switch(f){case
247:bm(CW);break;case
248:var
c=g9(a[2],b[2]);if(c!=0)return c|0;break;case
249:bm(CW);break;case
250:bm("equal: got Forward_tag, should not happen");break;case
251:bm("equal: abstract value");break;case
252:if(a!==b){var
c=S1(a,b);if(c!=0)return c|0}break;case
253:bm("equal: got Double_tag, should not happen");break;case
254:bm("equal: got Double_array_tag, should not happen");break;case
255:bm("equal: got Custom_tag, should not happen");break;case
1247:bm(CW);break;case
1255:var
i=Gs(a);if(i!=Gs(b))return a.caml_custom<b.caml_custom?-1:1;if(!i)bm("compare: abstract value");var
c=i(a,b,d);if(c!=c)return d?-1:c;if(c!==(c|0))return-1;if(c!=0)return c|0;break;case
1256:var
c=a.compare(b,d);if(c!=c)return d?-1:c;if(c!==(c|0))return-1;if(c!=0)return c|0;break;case
1000:a=+a;b=+b;if(a<b)return-1;if(a>b)return 1;if(a!=b){if(!d)return NaN;if(a==a)return 1;if(b==b)return-1}break;case
1001:if(a<b)return-1;if(a>b)return 1;if(a!=b){if(!d)return NaN;if(a==a)return 1;if(b==b)return-1}break;case
1251:if(a!==b){if(!d)return NaN;return 1}break;case
1252:var
a=hw(a),b=hw(b);if(a!==b){if(a<b)return-1;if(a>b)return 1}break;case
12520:var
a=a.toString(),b=b.toString();if(a!==b){if(a<b)return-1;if(a>b)return 1}break;case
246:case
254:default:if(a.length!=b.length)return a.length<b.length?-1:1;if(a.length>1)e.push(a,b,1);break}}if(e.length==0)return 0;var
h=e.pop();b=e.pop();a=e.pop();if(h+1<a.length)e.push(a,b,h+1);a=a[h];b=b[h]}}function
S3(a,b){return q6(a,b,true)}function
bGM(){return[0]}function
aW(a){if(a<0)bm("Bytes.create");return new
g6(a?2:9,i,a)}function
S7(a,b){return+(q6(a,b,false)==0)}function
bGN(a,c,b,d){if(b>0)if(c==0&&(b>=a.l||a.t==2&&b>=a.c.length))if(d==0){a.c=i;a.t=2}else{a.c=jp(b,String.fromCharCode(d));a.t=b==a.l?0:2}else{if(a.t!=4)q7(a);for(b+=c;c<b;c++)a.c[c]=d}return 0}function
kL(a){var
b;a=hw(a);b=+a;if(a.length>0&&b===b)return b;a=a.replace(/_/g,i);b=+a;if(a.length>0&&b===b||/^[+-]?nan$/i.test(a))return b;var
c=/^ *([+-]?)0x([0-9a-f]+)\.?([0-9a-f]*)(p([+-]?[0-9]+))?/i.exec(a);if(c){var
d=c[3].replace(/0+$/,i),f=parseInt(c[1]+c[2]+d,16),e=(c[5]|0)-4*d.length;b=f*Math.pow(2,e);return b}if(/^\+?inf(inity)?$/i.test(a))return Infinity;if(/^-inf(inity)?$/i.test(a))return-Infinity;bZ("float_of_string")}function
GE(d){d=hw(d);var
e=d.length;if(e>31)bm("format_int: format too long");var
a={justify:ho,signstyle:gk,filler:cN,alternate:false,base:0,signedconv:false,width:0,uppercase:false,sign:1,prec:-1,conv:R_};for(var
c=0;c<e;c++){var
b=d.charAt(c);switch(b){case"-":a.justify=gk;break;case"+":case" ":a.signstyle=b;break;case"0":a.filler=bV;break;case"#":a.alternate=true;break;case"1":case"2":case"3":case"4":case"5":case"6":case"7":case"8":case"9":a.width=0;while(b=d.charCodeAt(c)-48,b>=0&&b<=9){a.width=a.width*10+b;c++}c--;break;case".":a.prec=0;c++;while(b=d.charCodeAt(c)-48,b>=0&&b<=9){a.prec=a.prec*10+b;c++}c--;case"d":case"i":a.signedconv=true;case"u":a.base=10;break;case"x":a.base=16;break;case"X":a.base=16;a.uppercase=true;break;case"o":a.base=8;break;case"e":case"f":case"g":a.signedconv=true;a.conv=b;break;case"E":case"F":case"G":a.signedconv=true;a.uppercase=true;a.conv=b.toLowerCase();break}}return a}function
Gt(b,f){if(b.uppercase)f=f.toUpperCase();var
e=f.length;if(b.signedconv&&(b.sign<0||b.signstyle!=gk))e++;if(b.alternate){if(b.base==8)e+=1;if(b.base==16)e+=2}var
c=i;if(b.justify==ho&&b.filler==cN)for(var
d=e;d<b.width;d++)c+=cN;if(b.signedconv)if(b.sign<0)c+=gk;else
if(b.signstyle!=gk)c+=b.signstyle;if(b.alternate&&b.base==8)c+=bV;if(b.alternate&&b.base==16)c+=im;if(b.justify==ho&&b.filler==bV)for(var
d=e;d<b.width;d++)c+=bV;c+=f;if(b.justify==gk)for(var
d=e;d<b.width;d++)c+=cN;return a(c)}function
Gu(i,c){function
j(a,b){if(Math.abs(a)<1.0)return a.toFixed(b);else{var
c=parseInt(a.toString().split(ho)[1]);if(c>20){c-=20;a/=Math.pow(10,c);a+=new
Array(c+1).join(bV);if(b>0)a=a+fr+new
Array(b+1).join(bV);return a}else
return a.toFixed(b)}}var
a,e=GE(i),d=e.prec<0?6:e.prec;if(c<0||c==0&&1/c==-Infinity){e.sign=-1;c=-c}if(isNaN(c)){a=Ct;e.filler=cN}else
if(!isFinite(c)){a="inf";e.filler=cN}else
switch(e.conv){case"e":var
a=c.toExponential(d),b=a.length;if(a.charAt(b-3)==Ge)a=a.slice(0,b-1)+bV+a.slice(b-1);break;case"f":a=j(c,d);break;case"g":d=d?d:1;a=c.toExponential(d-1);var
h=a.indexOf(Ge),g=+a.slice(h+1);if(g<-4||c>=1e21||c.toFixed(0).length>d){var
b=h-1;while(a.charAt(b)==bV)b--;if(a.charAt(b)==fr)b--;a=a.slice(0,b+1)+a.slice(h);b=a.length;if(a.charAt(b-3)==Ge)a=a.slice(0,b-1)+bV+a.slice(b-1);break}else{var
f=d;if(g<0){f-=g+1;a=c.toFixed(f)}else
while(a=c.toFixed(f),a.length>d+1)f--;if(f){var
b=a.length-1;while(a.charAt(b)==bV)b--;if(a.charAt(b)==fr)b--;a=a.slice(0,b+1)}}break}return Gt(e,a)}function
q8(e,c){if(hw(e)==ks)return a(i+c);var
b=GE(e);if(c<0)if(b.signedconv){b.sign=-1;c=-c}else
c>>>=0;var
d=c.toString(b.base);if(b.prec>=0){b.filler=cN;var
f=b.prec-d.length;if(f>0)d=jp(f,bV)+d}return Gt(b,d)}var
Tg=0;function
hu(){return Tg++}function
S9(){return[0]}var
rb=[];function
ag(i,h,d){var
b=i[1],g=rb[d];if(g===undefined)for(var
f=rb.length;f<d;f++)rb[f]=0;else
if(b[g]===h)return b[g-1];var
a=3,e=b[1]*2+1,c;while(a<e){c=a+e>>1|1;if(h<b[c+1])e=c-2;else
a=c}rb[d]=a+1;return h==b[a+1]?b[a]:0}function
bHl(e){for(var
f=i,b=f,a,h,c=0,g=e.length;c<g;c++){a=e.charCodeAt(c);if(a<cW){for(var
d=c+1;d<g&&(a=e.charCodeAt(d))<cW;d++);if(d-c>yB){b.substr(0,1);f+=b;b=i;f+=e.slice(c,d)}else
b+=e.slice(c,d);if(d==g)break;c=d}if(a<Rv){b+=String.fromCharCode(0xc0|a>>6);b+=String.fromCharCode(cW|a&iX)}else
if(a<0xd800||a>=Sv)b+=String.fromCharCode(Rc|a>>12,cW|a>>6&iX,cW|a&iX);else
if(a>=0xdbff||c+1==g||(h=e.charCodeAt(c+1))<RM||h>Sv)b+="\xef\xbf\xbd";else{c++;a=(a<<10)+h-0x35fdc00;b+=String.fromCharCode(Qk|a>>18,cW|a>>12&iX,cW|a>>6&iX,cW|a&iX)}if(b.length>kB){b.substr(0,1);f+=b;b=i}}return f+b}function
bGK(a){var
b=9;if(!To(a))b=8,a=bHl(a);return new
g6(b,a,a.length)}function
cs(a){return bGK(a)}function
bGQ(a,c,l){if(!isFinite(a)){if(isNaN(a))return cs(Ct);return cs(a>0?SN:"-infinity")}var
j=a==0&&1/a==-Infinity?1:a>=0?0:1;if(j)a=-a;var
d=0;if(a==0);else
if(a<1)while(a<1&&d>-1022){a*=2;d--}else
while(a>=2){a/=2;d++}var
k=d<0?i:ho,e=i;if(j)e=gk;else
switch(l){case
43:e=ho;break;case
32:e=cN;break;default:break}if(c>=0&&c<13){var
g=Math.pow(2,c*4);a=Math.round(a*g)/g}var
b=a.toString(16);if(c>=0){var
h=b.indexOf(fr);if(h<0)b+=fr+jp(c,bV);else{var
f=h+1+c;if(b.length<f)b+=jp(f-b.length,bV);else
b=b.substr(0,f)}}return cs(e+im+b+"p"+k+d.toString(10))}function
bGY(a){return+a.isZero()}function
q$(a){return new
aj(a&fH,a>>24&fH,a>>31&cV)}function
bG1(a){return a.toInt()}function
bGX(a){return+a.isNeg()}function
Ta(a){return a.neg()}function
bGV(g,c){var
a=GE(g);if(a.signedconv&&bGX(c)){a.sign=-1;c=Ta(c)}var
b=i,h=q$(a.base),f="0123456789abcdef";do{var
e=c.udivmod(h);c=e.quotient;b=f.charAt(bG1(e.modulus))+b}while(!bGY(c));if(a.prec>=0){a.filler=cN;var
d=a.prec-b.length;if(d>0)b=jp(d,bV)+b}return Gt(a,b)}function
b9(a){return a.l}function
ah(a){return b9(a)}function
K(b,a){return bM(b,a)}function
bGS(a,b){return a.add(b)}function
bG0(a,b){return a.mul(b)}function
GA(a,b){return a.ucompare(b)<0}function
Th(c){var
a=0,e=ah(c),b=10,d=1;if(e>0)switch(K(c,a)){case
45:a++;d=-1;break;case
43:a++;d=1;break}if(a+1<e&&K(c,a)==48)switch(K(c,a+1)){case
120:case
88:b=16;a+=2;break;case
111:case
79:b=8;a+=2;break;case
98:case
66:b=2;a+=2;break;case
117:case
85:a+=2;break}return[a,d,b]}function
rc(a){if(a>=48&&a<=57)return a-48;if(a>=65&&a<=90)return a-55;if(a>=97&&a<=aO)return a-87;return-1}function
jm(f){var
g=Th(f),e=g[0],i=g[1],c=g[2],h=q$(c),j=new
aj(fH,0xfffffff,cV).udivmod(h).quotient,d=K(f,e),a=rc(d);if(a<0||a>=c)bZ(hk);var
b=q$(a);for(;;){e++;d=K(f,e);if(d==95)continue;a=rc(d);if(a<0||a>=c)break;if(GA(j,b))bZ(hk);a=q$(a);b=bGS(bG0(h,b),a);if(GA(b,a))bZ(hk)}if(e!=ah(f))bZ(hk);if(c==10&&GA(new
aj(0,0,cY),b))bZ(hk);if(i<0)b=Ta(b);return b}function
jn(a){return a.toFloat()}function
c6(f){var
h=Th(f),c=h[0],i=h[1],d=h[2],g=ah(f),j=-1>>>0,e=c<g?K(f,c):0,b=rc(e);if(b<0||b>=d)bZ(hk);var
a=b;for(c++;c<g;c++){e=K(f,c);if(e==95)continue;b=rc(e);if(b<0||b>=d)break;a=d*a+b;if(a>j)bZ(hk)}if(c!=g)bZ(hk);a=i*a;if(d==10&&(a|0)!=a)bZ(hk);return a|0}function
bG4(a){return a.slice(1)}function
bG5(a){return!!a}function
am(a){return a.toUtf16()}function
bG6(b){var
c={};for(var
a=1;a<b.length;a++){var
d=b[a];c[am(d[1])]=d[2]}return c}function
bG7(a,b){return+(q6(a,b,false)<0)}function
jo(a,d){if(a<0)kJ();var
a=a+1|0,b=new
Array(a);b[0]=0;for(var
c=1;c<a;c++)b[c]=d;return b}function
bG9(){return 0}function
aK(a){GF(aX.Sys_error,a)}var
g_=new
Array();function
iu(c){var
a=g_[c];if(!a.opened)aK("Cannot flush a closed channel");if(!a.buffer||a.buffer==i)return 0;if(a.fd&&aX.fds[a.fd]&&aX.fds[a.fd].output){var
b=aX.fds[a.fd].output;switch(b.length){case
2:b(c,a.buffer);break;default:b(a.buffer)}}a.buffer=i;return 0}function
kP(){return typeof
ar.process!=="undefined"&&typeof
ar.process.versions!=="undefined"&&typeof
ar.process.versions.node!=="undefined"}function
bHq(){function
a(a){if(a.charAt(0)===bT)return[i,a.substring(1)];return}function
b(c){var
g=/^([a-zA-Z]:|[\\/]{2}[^\\/]+[\\/]+[^\\/]+)?([\\/])?([\s\S]*?)$/,a=g.exec(c),b=a[1]||i,e=Boolean(b&&b.charAt(1)!==ig);if(Boolean(a[2]||e)){var
d=a[1]||i,f=a[2]||i;return[d,c.substring(d.length+f.length)]}return}return kP()&&ar.process&&ar.process.platform?ar.process.platform===Q3?b:a:a}var
GK=bHq();function
Tn(a){return a.slice(-1)!==bT?a+bT:a}if(kP()&&ar.process&&ar.process.cwd)var
kK=ar.process.cwd().replace(/\\/g,bT);else
var
kK="/static";kK=Tn(kK);function
bG8(a){a=am(a);if(!GK(a))a=kK+a;var
e=GK(a),d=e[1].split(bT),b=[];for(var
c=0;c<d.length;c++)switch(d[c]){case"..":if(b.length>1)b.pop();break;case".":break;default:b.push(d[c]);break}b.unshift(e[0]);b.orig=a;return b}var
bHt=["E2BIG","EACCES","EAGAIN","EBADF","EBUSY","ECHILD","EDEADLK","EDOM",Q6,"EFAULT","EFBIG","EINTR","EINVAL","EIO","EISDIR","EMFILE","EMLINK","ENAMETOOLONG","ENFILE","ENODEV",F2,"ENOEXEC","ENOLCK","ENOMEM","ENOSPC","ENOSYS",Fn,SJ,"ENOTTY","ENXIO","EPERM","EPIPE","ERANGE","EROFS","ESPIPE","ESRCH","EXDEV","EWOULDBLOCK","EINPROGRESS","EALREADY","ENOTSOCK","EDESTADDRREQ","EMSGSIZE","EPROTOTYPE","ENOPROTOOPT","EPROTONOSUPPORT","ESOCKTNOSUPPORT","EOPNOTSUPP","EPFNOSUPPORT","EAFNOSUPPORT","EADDRINUSE","EADDRNOTAVAIL","ENETDOWN","ENETUNREACH","ENETRESET","ECONNABORTED","ECONNRESET","ENOBUFS","EISCONN","ENOTCONN","ESHUTDOWN","ETOOMANYREFS","ETIMEDOUT","ECONNREFUSED","EHOSTDOWN","EHOSTUNREACH","ELOOP","EOVERFLOW"];function
iy(d,f,e,a){var
b=bHt.indexOf(d);if(b<0){if(a==null)a=-9999;b=[0,a]}var
c=[b,cs(f||i),cs(e||i)];return c}var
Tf={};function
iv(a){return Tf[a]}function
ix(b,a){throw[0,b].concat(a)}function
bGJ(a){return new
g6(4,a,a.length)}function
bHd(a){aK(a+qc)}function
bGI(b,a){if(a>>>0>=b.l)S0();return bM(b,a)}function
ST(){}function
bt(a){this.data=a}bt.prototype=new
ST();bt.prototype.truncate=function(a){var
b=this.data;this.data=aW(a|0);hS(b,0,this.data,0,a)};bt.prototype.length=function(){return b9(this.data)};bt.prototype.write=function(b,d,g,a){var
c=this.length();if(b+a>=c){var
e=aW(b+a),f=this.data;this.data=e;hS(f,0,this.data,0,c)}hT(d,g,this.data,b,a);return 0};bt.prototype.read=function(c,a,d,b){var
e=this.length();hS(this.data,c,a,d,b);return 0};bt.prototype.read_one=function(a){return bGI(this.data,a)};bt.prototype.close=function(){};bt.prototype.constructor=bt;function
bY(b,a){this.content={};this.root=b;this.lookupFun=a}bY.prototype.nm=function(a){return this.root+a};bY.prototype.create_dir_if_needed=function(d){var
c=d.split(bT),b=i;for(var
a=0;a<c.length-1;a++){b+=c[a]+bT;if(this.content[b])continue;this.content[b]=Symbol("directory")}};bY.prototype.slash=function(a){return/\/$/.test(a)?a:a+bT};bY.prototype.lookup=function(b){if(!this.content[b]&&this.lookupFun){var
c=this.lookupFun(a(this.root),a(b));if(c!==0){this.create_dir_if_needed(b);this.content[b]=new
bt(g8(c[1]))}}};bY.prototype.exists=function(a){if(a==i)return 1;var
b=this.slash(a);if(this.content[b])return 1;this.lookup(a);return this.content[a]?1:0};bY.prototype.mkdir=function(c,e,d){var
b=d&&iv(FW);if(this.exists(c))if(b)ix(b,iy(Q6,AS,this.nm(c)));else
aK(c+": File exists");var
a=/^(.*)\/[^/]+/.exec(c);a=a&&a[1]||i;if(!this.exists(a))if(b)ix(b,iy(F2,AS,this.nm(a)));else
aK(a+qc);if(!this.is_dir(a))if(b)ix(b,iy(Fn,AS,this.nm(a)));else
aK(a+zB);this.create_dir_if_needed(this.slash(c))};bY.prototype.rmdir=function(a,f){var
b=f&&iv(FW),c=a==i?i:this.slash(a),e=new
RegExp(oG+c+Pw);if(!this.exists(a))if(b)ix(b,iy(F2,xE,this.nm(a)));else
aK(a+qc);if(!this.is_dir(a))if(b)ix(b,iy(Fn,xE,this.nm(a)));else
aK(a+zB);for(var
d
in
this.content)if(d.match(e))if(b)ix(b,iy(SJ,xE,this.nm(a)));else
aK(this.nm(a)+": Directory not empty");delete
this.content[c]};bY.prototype.readdir=function(a){var
f=a==i?i:this.slash(a);if(!this.exists(a))aK(a+qc);if(!this.is_dir(a))aK(a+zB);var
g=new
RegExp(oG+f+Pw),d={},c=[];for(var
e
in
this.content){var
b=e.match(g);if(b&&!d[b[1]]){d[b[1]]=true;c.push(b[1])}}return c};bY.prototype.is_dir=function(a){if(a==i)return true;var
b=this.slash(a);return this.content[b]?1:0};bY.prototype.unlink=function(a){var
b=this.content[a]?true:false;delete
this.content[a];return b};bY.prototype.open=function(a,b){if(b.rdonly&&b.wronly)aK(this.nm(a)+Ri);if(b.text&&b.binary)aK(this.nm(a)+Sa);this.lookup(a);if(this.content[a]){if(this.is_dir(a))aK(this.nm(a)+" : is a directory");if(b.create&&b.excl)aK(this.nm(a)+Rh);var
c=this.content[a];if(b.truncate)c.truncate();return c}else
if(b.create){this.create_dir_if_needed(a);this.content[a]=new
bt(aW(0));return this.content[a]}else
bHd(this.nm(a))};bY.prototype.register=function(c,a){var
b;if(this.content[c])aK(this.nm(c)+Rh);if(ra(a))b=new
bt(a);if(GB(a))b=new
bt(g8(a));else
if(a
instanceof
Array)b=new
bt(bGJ(a));else
if(typeof
a==="string")b=new
bt(S2(a));else
if(a.toString){var
d=g8(cs(a.toString()));b=new
bt(d)}if(b){this.create_dir_if_needed(c);this.content[c]=b}else
aK(this.nm(c)+" : registering file with invalid content type")};bY.prototype.constructor=bY;function
bGD(d){var
c=ah(d),b=new
Array(c),a=0;for(;a<c;a++)b[a]=K(d,a);return b}function
bGC(a){if(a.t!=4)q7(a);return a.c}function
g7(a){this.fs=require("fs");this.fd=a}g7.prototype=new
ST();g7.prototype.truncate=function(a){try{this.fs.ftruncateSync(this.fd,a|0)}catch(a){aK(a.toString())}};g7.prototype.length=function(){try{return this.fs.fstatSync(this.fd).size}catch(a){aK(a.toString())}};g7.prototype.write=function(f,b,c,e){var
a=bGD(b);if(!(a
instanceof
ar.Uint8Array))a=new(ar.Uint8Array)(a);var
d=ar.Buffer.from(a);try{this.fs.writeSync(this.fd,d,c,e,f)}catch(a){aK(a.toString())}return 0};g7.prototype.read=function(g,d,c,f){var
a=bGC(d);if(!(a
instanceof
ar.Uint8Array))a=new(ar.Uint8Array)(a);var
e=ar.Buffer.from(a);try{this.fs.readSync(this.fd,e,c,f,g)}catch(a){aK(a.toString())}for(var
b=0;b<f;b++)it(d,c+b,e[c+b]);return 0};g7.prototype.read_one=function(c){var
b=new(ar.Uint8Array)(1),a=ar.Buffer.from(b);try{this.fs.readSync(this.fd,a,0,1,c)}catch(a){aK(a.toString())}return a[0]};g7.prototype.close=function(){try{this.fs.closeSync(this.fd)}catch(a){aK(a.toString())}};g7.prototype.constructor=g7;function
bu(a){this.fs=require("fs");this.root=a}bu.prototype.nm=function(a){return this.root+a};bu.prototype.exists=function(a){try{return this.fs.existsSync(this.nm(a))?1:0}catch(a){return 0}};bu.prototype.mkdir=function(b,a,c){try{this.fs.mkdirSync(this.nm(b),{mode:a});return 0}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.rmdir=function(a,b){try{this.fs.rmdirSync(this.nm(a));return 0}catch(a){this.raise_nodejs_error(a,b)}};bu.prototype.readdir=function(a,b){try{return this.fs.readdirSync(this.nm(a))}catch(a){this.raise_nodejs_error(a,b)}};bu.prototype.is_dir=function(a){try{return this.fs.statSync(this.nm(a)).isDirectory()?1:0}catch(a){aK(a.toString())}};bu.prototype.unlink=function(a,c){try{var
b=this.fs.existsSync(this.nm(a))?1:0;this.fs.unlinkSync(this.nm(a));return b}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.open=function(f,c,g){var
a=require("constants"),b=0;for(var
e
in
c)switch(e){case"rdonly":b|=a.O_RDONLY;break;case"wronly":b|=a.O_WRONLY;break;case"append":b|=a.O_WRONLY|a.O_APPEND;break;case"create":b|=a.O_CREAT;break;case"truncate":b|=a.O_TRUNC;break;case"excl":b|=a.O_EXCL;break;case"binary":b|=a.O_BINARY;break;case"text":b|=a.O_TEXT;break;case"nonblock":b|=a.O_NONBLOCK;break}try{var
d=this.fs.openSync(this.nm(f),b);return new
g7(d)}catch(a){this.raise_nodejs_error(a,g)}};bu.prototype.rename=function(b,a,c){try{this.fs.renameSync(this.nm(b),this.nm(a))}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.stat=function(b,c){try{var
a=this.fs.statSync(this.nm(b));return this.stats_from_js(a)}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.lstat=function(b,c){try{var
a=this.fs.lstatSync(this.nm(b));return this.stats_from_js(a)}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.symlink=function(d,c,a,b){try{this.fs.symlinkSync(this.nm(c),this.nm(a),d?"dir":"file");return 0}catch(a){this.raise_nodejs_error(a,b)}};bu.prototype.readlink=function(b,c){try{var
a=this.fs.readlinkSync(this.nm(b),Pi);return cs(a)}catch(a){this.raise_nodejs_error(a,c)}};bu.prototype.raise_nodejs_error=function(a,d){var
b=iv(FW);if(d&&b){var
c=iy(a.code,a.syscall,a.path,a.errno);ix(b,c)}else
aK(a.toString())};bu.prototype.stats_from_js=function(a){var
b;if(a.isFile())b=0;else
if(a.isDirectory())b=1;else
if(a.isCharacterDevice())b=2;else
if(a.isBlockDevice())b=3;else
if(a.isSymbolicLink())b=4;else
if(a.isFIFO())b=5;else
if(a.isSocket())b=6;return[0,a.dev,a.ino,b,a.mode,a.nlink,a.uid,a.gid,a.rdev,a.size,a.atimeMs,a.mtimeMs,a.ctimeMs]};bu.prototype.constructor=bu;function
S_(b){var
a=GK(b);if(!a)return;return a[0]+bT}var
rd=S_(kK)||bZ("unable to compute caml_root"),jq=[];if(kP())jq.push({path:rd,device:new
bu(rd)});else
jq.push({path:rd,device:new
bY(rd)});jq.push({path:RT,device:new
bY(RT)});function
bHs(b){var
g=bG8(b),b=g.join(bT),f=Tn(b),c;for(var
e=0;e<jq.length;e++){var
a=jq[e];if(f.search(a.path)==0&&(!c||c.path.length<a.path.length))c={path:a.path,device:a.device,rest:b.substring(a.path.length,b.length)}}if(!c&&kP()){var
d=S_(b);if(d&&d.match(/^[a-zA-Z]:\/$/)){var
a={path:d,device:new
bu(d)};jq.push(a);c={path:a.path,device:a.device,rest:b.substring(a.path.length,b.length)}}}if(c)return c;aK("no device found for "+f)}function
Tl(e,f){var
b=g_[e],d=a(f),c=ah(d);b.file.write(b.offset,d,0,c);b.offset+=c;return 0}function
bHm(a){var
a=GJ(a),b=ar;if(b.process&&b.process.stdout&&b.process.stdout.write)b.process.stderr.write(a);else{if(a.charCodeAt(a.length-1)==10)a=a.substr(0,a.length-1);var
c=b.console;c&&c.error&&c.error(a)}}function
bHn(a){var
a=GJ(a),b=ar;if(b.process&&b.process.stdout&&b.process.stdout.write)b.process.stdout.write(a);else{if(a.charCodeAt(a.length-1)==10)a=a.substr(0,a.length-1);var
c=b.console;c&&c.log&&c.log(a)}}function
re(c,e,d,a){if(aX.fds===undefined)aX.fds=new
Array();a=a?a:{};var
b={};b.file=d;b.offset=a.append?d.length():0;b.flags=a;b.output=e;aX.fds[c]=b;if(!aX.fd_last_idx||c>aX.fd_last_idx)aX.fd_last_idx=c;return c}function
bHu(c,b,g){var
a={};while(b){switch(b[1]){case
0:a
